File: packages/indexer/indexer/worker-protocol/native-static.ts
/**
 * Native static compiler protocol contracts.
 *
 * These schemas describe the internal Go/Rust/TypeScript handoff that will
 * replace the syntax-record bridge. They are intentionally fact- and
 * evidence-shaped: the protocol carries JSON-safe Crux data, never TypeScript
 * or Oxc AST objects.
 *
 * @module
 */

import { z } from 'zod'
import { nativeStaticParserInterestFields } from './native-static-parser-interests'

/** Current native static compiler protocol version. */
export const NATIVE_STATIC_COMPILER_PROTOCOL_VERSION = 1 as const

/** Native static compiler method names used on the JSON-lines boundary. */
export type NativeStaticCompilerMethod =
  | 'nativeStaticPrepare'
  | 'nativeStaticAnalyze'
  | 'nativeStaticFinalize'
  | 'nativeStaticCompile'

const identityComponentSchema = z
  .object({
    name: z.string().min(1),
    version: z.string().min(1),
    digest: z.string().min(1).optional(),
  })
  .strict()

const nonNegativeNumber = z.number().finite().nonnegative()
const unknownArraySchema = z.array(z.unknown())

/**
 * Versioned identity for every compiler input that can affect native static
 * output or cache reuse.
 */
export const NativeStaticRunIdentitySchema = z
  .object({
    protocolVersion: z.literal(NATIVE_STATIC_COMPILER_PROTOCOL_VERSION),
    compiler: identityComponentSchema,
    oxc: identityComponentSchema,
    primitiveManifest: identityComponentSchema,
    relationPolicy: identityComponentSchema,
    extensionManifests: z.array(identityComponentSchema),
    firstPartyGraphRules: identityComponentSchema,
    compilerProjection: identityComponentSchema,
  })
  .strict()

/** Native static compiler run identity. */
export type NativeStaticRunIdentity = z.infer<typeof NativeStaticRunIdentitySchema>

/**
 * Machine-readable telemetry emitted by each native static stage.
 *
 * Go benchmarks and devtools diagnostics use this to explain whether Node was
 * required, why native-only eligibility failed, and how many files/facts/cache
 * entries moved through the stage.
 */
export const NativeStaticTelemetrySchema = z
  .object({
    node: z
      .object({
        started: z.boolean(),
        reasons: z.array(z.string()),
      })
      .strict(),
    nativeOnly: z
      .object({
        eligible: z.boolean(),
        reasons: z.array(z.string()),
      })
      .strict(),
    timings: z.array(
      z
        .object({
          name: z.string().min(1),
          durationMs: nonNegativeNumber,
          count: nonNegativeNumber.optional(),
        })
        .strict(),
    ),
    files: z
      .object({
        selected: nonNegativeNumber,
        cacheHits: nonNegativeNumber,
        cacheMisses: nonNegativeNumber,
        analyzed: nonNegativeNumber,
        skipped: nonNegativeNumber,
      })
      .strict(),
    cache: z
      .object({
        readHits: nonNegativeNumber,
        readMisses: nonNegativeNumber,
        writes: nonNegativeNumber,
        writeErrors: nonNegativeNumber,
      })
      .strict(),
    facts: z
      .object({
        definitions: nonNegativeNumber,
        relations: nonNegativeNumber,
        sourceRefs: nonNegativeNumber,
        diagnostics: nonNegativeNumber,
        lintFindings: nonNegativeNumber,
        ruleDescriptors: nonNegativeNumber,
        sources: nonNegativeNumber,
        sourceGraph: nonNegativeNumber,
      })
      .strict(),
  })
  .strict()

/** Native static stage telemetry. */
export type NativeStaticTelemetry = z.infer<typeof NativeStaticTelemetrySchema>

/** Source file identity selected for native static planning. */
export const NativeStaticSourceFileSchema = z
  .object({
    file: z.string().min(1),
    sourceHash: z.string().min(1),
    cacheKey: z.string().min(1).optional(),
  })
  .strict()

export type NativeStaticSourceFile = z.infer<typeof NativeStaticSourceFileSchema>

/** Source file selected for native static analysis. */
export const NativeStaticFileInputSchema = z
  .object({
    file: z.string().min(1),
    sourceHash: z.string().min(1),
    sourceText: z.string().optional(),
  })
  .strict()

export type NativeStaticFileInput = z.infer<typeof NativeStaticFileInputSchema>

/** Normalized native static source plan shared by prepare and analyze. */
export const NativeStaticPreparedPlanSchema = z
  .object({
    root: z.string().min(1),
    projectName: z.string().min(1).optional(),
    files: z.array(NativeStaticSourceFileSchema),
    /** Extraction owners inside the wider parse set. */
    primaryFiles: z.array(NativeStaticSourceFileSchema).optional(),
    cacheHits: z.array(NativeStaticSourceFileSchema),
    cacheMisses: z.array(NativeStaticSourceFileSchema),
    ...nativeStaticParserInterestFields,
  })
  .strict()

export type NativeStaticPreparedPlan = z.infer<typeof NativeStaticPreparedPlanSchema>

const nativeStaticRequestBase = {
  protocolVersion: z.literal(NATIVE_STATIC_COMPILER_PROTOCOL_VERSION),
  identity: NativeStaticRunIdentitySchema,
} as const

const nativeStaticResponseBase = {
  protocolVersion: z.literal(NATIVE_STATIC_COMPILER_PROTOCOL_VERSION),
  diagnostics: unknownArraySchema,
  telemetry: NativeStaticTelemetrySchema,
} as const

/** Request for the Rust compiler to normalize source/cache planning input. */
export const NativeStaticPrepareRequestSchema = z
  .object({
    ...nativeStaticRequestBase,
    method: z.literal('nativeStaticPrepare'),
    root: z.string().min(1),
    projectName: z.string().min(1).optional(),
    configPath: z.string().min(1).optional(),
    files: z.array(NativeStaticSourceFileSchema),
    /** Extraction owners inside the wider parse set. */
    primaryFiles: z.array(NativeStaticSourceFileSchema).optional(),
    ...nativeStaticParserInterestFields,
    cacheInputs: unknownArraySchema.optional(),
    extensionHost: z.unknown().optional(),
  })
  .strict()

/** Response from native static prepare. */
export const NativeStaticPrepareResponseSchema = z
  .object({
    ...nativeStaticResponseBase,
    method: z.literal('nativeStaticPrepare'),
    plan: NativeStaticPreparedPlanSchema,
  })
  .strict()

/** Request for the Rust compiler to parse selected cache misses and emit facts/evidence. */
export const NativeStaticAnalyzeRequestSchema = z
  .object({
    ...nativeStaticRequestBase,
    method: z.literal('nativeStaticAnalyze'),
    stream: z.literal(true),
    plan: NativeStaticPreparedPlanSchema,
    files: z.array(NativeStaticFileInputSchema),
    extensionEvidenceInterests: z.unknown().optional(),
  })
  .strict()

/** Response from native static analyze. */
export const NativeStaticAnalyzeResponseSchema = z
  .object({
    ...nativeStaticResponseBase,
    method: z.literal('nativeStaticAnalyze'),
    facts: unknownArraySchema,
    extensionEvidenceJobs: unknownArraySchema,
  })
  .strict()

/** Request for native relation/rule/cache finalization. */
export const NativeStaticFinalizeRequestSchema = z
  .object({
    ...nativeStaticRequestBase,
    method: z.literal('nativeStaticFinalize'),
    stream: z.literal(true).optional(),
    nativeFacts: unknownArraySchema,
    extensionFacts: unknownArraySchema,
    /** Post-merge facts used by native lint evaluation but not re-emitted. */
    lintFacts: unknownArraySchema.optional(),
    relationSpecs: z.unknown().optional(),
    ruleResults: z.unknown().optional(),
    lintConfig: z.unknown().optional(),
    lintFiles: z.array(z.string()).optional(),
    emitBuiltinLints: z.boolean().optional(),
    patchPhase: z.enum(['ast', 'semantic', 'runtime', 'quality']).optional(),
    patchInvalidates: z.unknown().optional(),
    cache: z.unknown().optional(),
  })
  .strict()

/** Response from native static finalization. */
export const NativeStaticFinalizeResponseSchema = z
  .object({
    protocolVersion: z.literal(NATIVE_STATIC_COMPILER_PROTOCOL_VERSION),
    method: z.literal('nativeStaticFinalize'),
    events: unknownArraySchema,
    telemetry: NativeStaticTelemetrySchema,
  })
  .strict()

/** Request for native-only parse, relation/rule finalization, and streamed patch events. */
export const NativeStaticCompileRequestSchema = z
  .object({
    ...nativeStaticRequestBase,
    method: z.literal('nativeStaticCompile'),
    stream: z.literal(true),
    plan: NativeStaticPreparedPlanSchema,
    files: z.array(NativeStaticFileInputSchema),
    nativeFacts: unknownArraySchema,
    extensionFacts: unknownArraySchema,
    relationSpecs: z.unknown().optional(),
    lintConfig: z.unknown().optional(),
    lintFiles: z.array(z.string()).optional(),
    emitBuiltinLints: z.boolean().optional(),
  })
  .strict()

/** Response from native-only streamed compilation. */
export const NativeStaticCompileResponseSchema = z
  .object({
    protocolVersion: z.literal(NATIVE_STATIC_COMPILER_PROTOCOL_VERSION),
    method: z.literal('nativeStaticCompile'),
    events: unknownArraySchema,
    telemetry: NativeStaticTelemetrySchema,
  })
  .strict()

/** Native static compiler request union. */
export const NativeStaticCompilerRequestSchema = z.discriminatedUnion('method', [
  NativeStaticPrepareRequestSchema,
  NativeStaticAnalyzeRequestSchema,
  NativeStaticFinalizeRequestSchema,
  NativeStaticCompileRequestSchema,
])

/** Native static compiler response union. */
export const NativeStaticCompilerResponseSchema = z.discriminatedUnion('method', [
  NativeStaticPrepareResponseSchema,
  NativeStaticAnalyzeResponseSchema,
  NativeStaticFinalizeResponseSchema,
  NativeStaticCompileResponseSchema,
])

export type NativeStaticCompilerRequest = z.infer<typeof NativeStaticCompilerRequestSchema>

export type NativeStaticCompilerResponse = z.infer<typeof NativeStaticCompilerResponseSchema>

File: crates/protocol/src/native_static.rs
//! Data-only Rust ABI shapes for the native static compiler protocol.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::{StaticSyntaxCallInterest, StaticSyntaxConstructorInterest};

/// Current Static Index compiler protocol version.
pub const STATIC_INDEX_PROTOCOL_VERSION: u8 = 1;

/// Method string for Static Index planning.
pub const STATIC_INDEX_PREPARE_METHOD: &str = "staticIndexPrepare";

/// Method string for Static Index file analysis.
pub const STATIC_INDEX_ANALYZE_METHOD: &str = "staticIndexAnalyze";

/// Method string for Static Index final patch/event materialization.
pub const STATIC_INDEX_FINALIZE_METHOD: &str = "staticIndexFinalize";

/// Method string for streamed Static Index compile.
pub const STATIC_INDEX_COMPILE_METHOD: &str = "staticIndexCompile";

/// Native static compiler method discriminator.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum NativeStaticMethod {
    #[serde(rename = "staticIndexPrepare")]
    Prepare,
    #[serde(rename = "staticIndexAnalyze")]
    Analyze,
    #[serde(rename = "staticIndexFinalize")]
    Finalize,
    #[serde(rename = "staticIndexCompile")]
    Compile,
}

/// Complete identity for a native static compiler run.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticRunIdentity {
    pub protocol_version: u8,
    pub compiler: NativeStaticVersionIdentity,
    pub oxc: NativeStaticVersionIdentity,
    pub primitive_manifest: NativeStaticDigestIdentity,
    pub relation_policy: NativeStaticDigestIdentity,
    pub extension_manifests: Vec<NativeStaticDigestIdentity>,
    pub rule_descriptors: NativeStaticDigestIdentity,
    pub compiler_projection: NativeStaticDigestIdentity,
}

/// Name/version identity for compiler-owned components.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticVersionIdentity {
    pub name: String,
    pub version: String,
}

/// Name/version identity with an optional digest for cache-sensitive inputs.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticDigestIdentity {
    pub name: String,
    pub version: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub digest: Option<String>,
}

/// Compiler-owned Static Index identity manifest before project extension input is added.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticIdentityManifest {
    pub protocol_version: u8,
    pub compiler: NativeStaticVersionIdentity,
    pub oxc_frontend: NativeStaticVersionIdentity,
    pub primitive_manifest: NativeStaticDigestIdentity,
    pub relation_policy: NativeStaticDigestIdentity,
    pub rule_descriptors: NativeStaticDigestIdentity,
    pub compiler_projection: NativeStaticDigestIdentity,
}

/// Cross-stage telemetry for one native static compiler run.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticTelemetry {
    pub node: NativeStaticNodeTelemetry,
    pub native_only: NativeStaticNativeOnlyTelemetry,
    pub timings: Vec<NativeStaticTiming>,
    pub files: NativeStaticFileTelemetry,
    pub cache: NativeStaticCacheTelemetry,
    pub facts: NativeStaticFactTelemetry,
}

/// Whether Node started, plus machine-readable reasons when it did.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticNodeTelemetry {
    pub started: bool,
    pub reasons: Vec<String>,
}

/// Whether the run was eligible for native-only execution.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticNativeOnlyTelemetry {
    pub eligible: bool,
    pub reasons: Vec<String>,
}

/// Named timing measurement. `count` is for batched work such as files or jobs.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticTiming {
    pub name: String,
    pub duration_ms: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub count: Option<u64>,
}

/// File counts reported by prepare/analyze/finalize.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticFileTelemetry {
    pub selected: u64,
    pub cache_hits: u64,
    pub cache_misses: u64,
    pub analyzed: u64,
    pub skipped: u64,
}

/// Static cache counters for the run.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticCacheTelemetry {
    pub read_hits: u64,
    pub read_misses: u64,
    pub writes: u64,
    pub write_errors: u64,
}

/// Project Index fact counters emitted by the run.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticFactTelemetry {
    pub definitions: u64,
    pub relations: u64,
    pub source_refs: u64,
    pub diagnostics: u64,
    pub lint_findings: u64,
    pub rule_descriptors: u64,
    pub sources: u64,
    pub source_graph: u64,
}

/// Selected source file identity used by prepare and prepared plans.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticSourceFile {
    pub file: String,
    pub source_hash: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cache_key: Option<String>,
}

/// Prepared source plan passed from prepare into analyze.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticPlan {
    pub root: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_name: Option<String>,
    pub files: Vec<NativeStaticSourceFile>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub primary_files: Option<Vec<NativeStaticSourceFile>>,
    pub cache_hits: Vec<NativeStaticSourceFile>,
    pub cache_misses: Vec<NativeStaticSourceFile>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub call_names: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub call_interests: Vec<StaticSyntaxCallInterest>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub constructor_names: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub constructor_interests: Vec<StaticSyntaxConstructorInterest>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub prune_native_fact_call_names: Vec<String>,
}

/// Analyze-stage file payload. `sourceText` is present when Rust cannot read.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticAnalyzeFile {
    pub file: String,
    pub source_hash: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_text: Option<String>,
}

/// `staticIndexPrepare` request.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticPrepareRequest {
    pub protocol_version: u8,
    pub method: NativeStaticMethod,
    pub root: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub config_path: Option<String>,
    pub identity: NativeStaticRunIdentity,
    pub files: Vec<NativeStaticSourceFile>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub primary_files: Option<Vec<NativeStaticSourceFile>>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub call_names: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub call_interests: Vec<StaticSyntaxCallInterest>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub constructor_names: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub constructor_interests: Vec<StaticSyntaxConstructorInterest>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub prune_native_fact_call_names: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cache_inputs: Option<Vec<Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extension_host: Option<Value>,
}

/// `staticIndexPrepare` response.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticPrepareResponse {
    pub protocol_version: u8,
    pub method: NativeStaticMethod,
    pub plan: NativeStaticPlan,
    pub diagnostics: Vec<Value>,
    pub telemetry: NativeStaticTelemetry,
}

/// `staticIndexAnalyze` request.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticAnalyzeRequest {
    pub protocol_version: u8,
    pub method: NativeStaticMethod,
    #[serde(default, skip_serializing_if = "is_false")]
    pub stream: bool,
    pub identity: NativeStaticRunIdentity,
    pub plan: NativeStaticPlan,
    pub files: Vec<NativeStaticAnalyzeFile>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extension_evidence_interests: Option<Value>,
}

fn is_false(value: &bool) -> bool {
    !*value
}

/// `staticIndexAnalyze` response.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticAnalyzeResponse {
    pub protocol_version: u8,
    pub method: NativeStaticMethod,
    pub facts: Vec<Value>,
    pub diagnostics: Vec<Value>,
    pub extension_evidence_jobs: Vec<Value>,
    pub telemetry: NativeStaticTelemetry,
}

/// `staticIndexCompile` request for native-only AST indexing.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticCompileRequest {
    pub protocol_version: u8,
    pub method: NativeStaticMethod,
    #[serde(default, skip_serializing_if = "is_false")]
    pub stream: bool,
    pub identity: NativeStaticRunIdentity,
    pub plan: NativeStaticPlan,
    pub files: Vec<NativeStaticAnalyzeFile>,
    pub native_facts: Vec<Value>,
    pub extension_facts: Vec<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relation_specs: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub lint_config: Option<Value>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub lint_files: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub emit_builtin_lints: Option<bool>,
}

/// `staticIndexFinalize` request.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticFinalizeRequest {
    pub protocol_version: u8,
    pub method: NativeStaticMethod,
    #[serde(default, skip_serializing_if = "is_false")]
    pub stream: bool,
    pub identity: NativeStaticRunIdentity,
    pub native_facts: Vec<Value>,
    pub extension_facts: Vec<Value>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub lint_facts: Vec<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relation_specs: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rule_results: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub lint_config: Option<Value>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub lint_files: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub emit_builtin_lints: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub patch_phase: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub patch_invalidates: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cache: Option<Value>,
}

/// `staticIndexFinalize` response.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeStaticFinalizeResponse {
    pub protocol_version: u8,
    pub method: NativeStaticMethod,
    pub events: Vec<Value>,
    pub telemetry: NativeStaticTelemetry,
}
